orchagent/bufferorch.h

```diff
@@ -30,6 +31,68 @@ const string buffer_value_both              = "both";
 const string buffer_profile_list_field_name = "profile_list";
 const string buffer_headroom_type_field_name= "headroom_type";
 
+struct PortBufferProfileListTask
+{
+    struct PortContext
+    {
+        std::string port_name;
+        sai_object_id_t port_oid = SAI_NULL_OBJECT_ID;
+        SaiAttrWrapper attr = {};
+        sai_status_t status = SAI_STATUS_NOT_EXECUTED;
+    };
+
+    KeyOpFieldsValuesTuple kofvs;
+    std::vector<PortContext> ports;
+};
+
+struct PriorityGroupTask
+{
+    struct PgContext
+    {
+        size_t index;
+        bool update_sai = true;
+        bool counter_was_added = false;
+        bool counter_needs_to_add = false;
+        sai_object_id_t pg_id = SAI_NULL_OBJECT_ID;
+        SaiAttrWrapper attr = {};
+        sai_status_t status = SAI_STATUS_NOT_EXECUTED;
+    };
+
+    struct PortContext
+    {
+        std::string port_name;
+        std::vector<PgContext> pgs;
```

Comment:
so, the benefit of organizing the queues and PGs hierarchically is to avoid storing port-level variables for each queue?

Reply:
@stephenxs Inherently yes, port level context is shared for multiple queues of that port. 
So, if I understand correctly the task could look like this EthernetX,EthernetY,EthernetZ|queue1,queue2,queue3 and the question is why storing queues 1-3 for each port and not port per each queue?

I find that most of config is one port - multiple pgs/queues, so this data structure is optimal for that usual case. 
The code that processes PGs and queues looked like this:

```
for p in port:
   for q in queues:
      ...
```

and not the other way around, and so the data structure is derived from the code structure. So that the corresponding *Post method does the same two for-loops.